### ISSUE TITLE
Donate Intermediate Proposal

### ISSUE THREAD

**Issue by ZYHowell** [alpa-projects/alpa#125]
A recent PR #119 fixes the memory leak caused by too aggressive invar donation, and as a result, now we only donate two kinds of tensors: 
1. Invars that are allowed to be donated. Typically, if a user sets "auto" as `donate_argnums`, we exactly donate the optimizer;
2. Gradient buffers. When #microbatch > 1, gradient accumulation is required, and we create corresponding inputs (accumulated gradients from the last microbatch) and outputs (accumulated gradients after this microbatch).

To explain what we are not donated(but is donatable), we introduce a concept of `main copy`. The main copy of a tensor locates on the mesh where it is defined. In contrast, the tensor copies sent to other meshes are `template copies`. Main copy is only for intermediate and output vars, but not for input vars, because the definitions of inputs depend only on dataloader. 
This definition is natural from our current runtime: when we need a tensor, we get it by a resharding task from the mesh that defines the tensor.

Hence, typically there are two kinds of donatable buffers:
1. A template copy of intermediates. It can be donated at any time;
2. A main copy of intermediate no longer used by other stages, in the same mesh or not.

We do not donate the above types for one consideration: If a buffer is donated, there is a corresponding input-output alias in `XlaComputation`. The alias influences two aspects: 
1. As we merge forward, backward and apply gradient together in an auto-sharding pass, we cannot set up an alias for an intermediate because it is not an input;

2. In auto-sharding, if there is an alias between input_i and output_i, they will share the same ShardingSpec. Hence, if we set too many donations, the auto-sharding will be restricted. A rough experiment about how much the auto-sharding's performance is influenced is listed later. 

Our experiment reveals two facts:
1. The time for `Jax3DPipeline` to do garbage collection is significant. 
Without any change in runnable, but only make the runtime skip the garbage collection process of the two kinds of donatable above, we accelerate at least 10~15%(from 0.69s to 0.60s, other cases observe more improvements). 

2.  On the other hand,  less-constrained auto-sharding is important for performance:
Although not design a precise experiment, we have the observation by: switch the donation of input optimizer, but donate it to gradients instead of corresponding output optimizer. 
If we switch off the donation, that is we suffer from our runtime's garbage collection, the cost is 0.69s. But if we switch on the donation, that is to skip our garbage collection but add restrictions to auto-sharding, the time is slow down to 0.85s.

the donation of intermediate is critical in order to reduce the garbage collection's time cost, but it should not limit auto-sharding. In conclusion, we donate the two kinds of buffers discussed above but do this only after the auto-sharding pass. At that point, we only donate those with **same shape and same sharding spec**. This misses some opportunities of donation, but the overall performance is expected to be better.

There are two problems that remain:
1. When the user, for some purpose, manually set to donate other inputs e.g. batch, the alias is created and will influence the auto-sharding heavily;
2. The cost of our runtime to release intermediates is formidabale, why we do not get likewise results in our 2D parallel or other systems?
